File: pages/cover_letter.py
"""
HirePilot AI – Cover Letter Generator Page
Generate professional, personalised cover letters with one click.
"""

from __future__ import annotations

import streamlit as st

from components.ui_components import info_box, render_footer, success_box
from services.watsonx_service import generate_cover_letter
from utils.helpers import section_header, generate_pdf_bytes


def render() -> None:
    """Render the Cover Letter Generator page."""

    section_header("Cover Letter Generator", "✉️")
    st.markdown(
        "<p style='color:#6f6f6f;font-size:0.9rem;margin-top:-0.5rem;margin-bottom:1.5rem;'>"
        "Generate a professional, tailored cover letter in seconds using IBM watsonx AI."
        "</p>",
        unsafe_allow_html=True,
    )

    form_col, result_col = st.columns([1, 1], gap="large")

    # ── Left: Input Form ───────────────────────────────────
    with form_col:
        st.markdown(
            '<div style="background:#f4f4f4;border:1px solid #e5e7eb;border-radius:4px;padding:1.5rem;">',
            unsafe_allow_html=True,
        )
        st.markdown("### ✏️ Letter Details")

        company = st.text_input(
            "🏢 Company Name",
            placeholder="e.g. Stripe, Google, IBM…",
            key="cl_company",
        )
        role = st.text_input(
            "🎯 Role / Position",
            placeholder="e.g. Senior Software Engineer",
            key="cl_role",
        )
        job_description = st.text_area(
            "📋 Job Description",
            placeholder="Paste the job description here…",
            height=200,
            key="cl_jd",
        )
        tone = st.select_slider(
            "📢 Letter Tone",
            options=["Formal", "Professional", "Enthusiastic", "Bold"],
            value="Professional",
            key="cl_tone",
        )
        include_resume = st.checkbox(
            "📎 Incorporate my uploaded resume",
            value=bool(st.session_state.get("resume_text")),
            key="cl_use_resume",
        )

        st.markdown("</div>", unsafe_allow_html=True)

        st.markdown("<div style='height:1rem'></div>", unsafe_allow_html=True)

        generate_clicked = st.button(
            "✨ Generate Cover Letter",
            use_container_width=True,
            key="cl_generate",
        )

    # ── Right: Generated Letter ────────────────────────────
    with result_col:
        if generate_clicked:
            if not company.strip() or not role.strip():
                st.error("❌ Please fill in the Company Name and Role fields.")
            else:
                resume_text = st.session_state.get("resume_text", "") if include_resume else ""
                with st.spinner("✍️ IBM watsonx AI is crafting your cover letter…"):
                    result = generate_cover_letter(
                        company=company,
                        role=role,
                        job_description=job_description,
                        resume_text=resume_text,
                    )
                    st.session_state["cover_letter_text"] = result.get("cover_letter", "")

        letter = st.session_state.get("cover_letter_text", "")

        if letter:
            success_box(f"Cover letter generated for {st.session_state.get('cl_company', 'your target company')}!")

            # Metadata
            word_count = len(letter.split())
            st.markdown(
                f"""
                <div style="display:flex;gap:1.5rem;font-size:0.78rem;color:#6f6f6f;
                            margin-bottom:0.75rem;">
                    <span>📝 {word_count} words</span>
                    <span>🎨 Professional tone</span>
                    <span>✅ ATS-optimised</span>
                </div>
                """,
                unsafe_allow_html=True,
            )

            # Letter display
            st.markdown(
                f'<div class="cover-letter-box">{letter}</div>',
                unsafe_allow_html=True,
            )

            st.markdown("<div style='height:0.75rem'></div>", unsafe_allow_html=True)

            # Action buttons
            act_col1, act_col2, act_col3 = st.columns(3)

            with act_col1:
                # Copy to clipboard via st.code
                if st.button("📋 Copy Letter", use_container_width=True, key="cl_copy"):
                    st.code(letter, language=None)

            with act_col2:
                # Download as text
                st.download_button(
                    label="💾 Download TXT",
                    data=letter.encode("utf-8"),
                    file_name=f"cover_letter_{st.session_state.get('cl_company','').replace(' ','_')}.txt",
                    mime="text/plain",
                    use_container_width=True,
                    key="cl_download_txt",
                )

            with act_col3:
                # Download as PDF
                pdf_bytes = generate_pdf_bytes(letter, title=f"Cover Letter – {role} at {company}")
                st.download_button(
                    label="📄 Download PDF",
                    data=pdf_bytes,
                    file_name=f"cover_letter_{company.replace(' ','_')}.pdf",
                    mime="application/pdf",
                    use_container_width=True,
                    key="cl_download_pdf",
                )

            # Edit letter
            with st.expander("✏️ Edit Letter", expanded=False):
                edited = st.text_area(
                    "Edit Cover Letter",
                    value=letter,
                    height=400,
                    key="cl_editor",
                    label_visibility="collapsed",
                )
                if st.button("💾 Save Edits", key="cl_save_edit"):
                    st.session_state["cover_letter_text"] = edited
                    st.success("Saved!")
                    st.rerun()

        else:
            # Empty state
            st.markdown(
                """
                <div style="text-align:center;padding:4rem 2rem;color:#6f6f6f;">
                    <div style="font-size:3rem;margin-bottom:1rem;">✉️</div>
                    <div style="font-size:1rem;font-weight:500;color:#161616;">
                        Your cover letter will appear here
                    </div>
                    <div style="font-size:0.875rem;margin-top:0.5rem;">
                        Fill in the form on the left and click Generate.
                    </div>
                </div>
                """,
                unsafe_allow_html=True,
            )

    # ── Tips ───────────────────────────────────────────────
    st.markdown("<hr class='ibm-divider'>", unsafe_allow_html=True)
    section_header("Cover Letter Tips", "💡")

    tip_cols = st.columns(3)
    tips = [
        ("🎯 Tailor Every Letter", "Customise each letter for the specific company and role. Generic letters have a 3× lower response rate."),
        ("📊 Use Numbers", "Quantify your achievements wherever possible. 'Improved performance by 40%' beats 'improved performance'."),
        ("✂️ Keep It Concise", "Aim for 3–4 paragraphs, 250–350 words. Hiring managers spend an average of 7 seconds on first read."),
    ]
    for col, (title, body) in zip(tip_cols, tips):
        with col:
            st.markdown(
                f"""
                <div style="background:#f4f4f4;border-radius:4px;padding:1rem 1.25rem;height:100%;">
                    <div style="font-weight:600;margin-bottom:0.4rem;">{title}</div>
                    <div style="font-size:0.82rem;color:#6f6f6f;">{body}</div>
                </div>
                """,
                unsafe_allow_html=True,
            )

    render_footer()

File: pages/cover-letter.js
// HirePilot AI – Premium Cover Letter Generator
import { useState } from 'react';
import Layout from '../components/Layout';
import { SectionHeader, Loading } from '../components/UI';

function buildMockLetter(company, role) {
  const today = new Date().toLocaleDateString('en-US', { year: 'numeric', month: 'long', day: '2-digit' });
  return `${today}

Hiring Manager
${company}

Dear Hiring Manager,

I am writing to express my strong interest in the ${role} position at ${company}. With over three years of hands-on experience building scalable web applications and a deep passion for crafting elegant, maintainable code, I am confident I would make an immediate and lasting contribution to your engineering team.

Throughout my career, I have delivered high-impact projects — reducing API response times by 40%, scaling distributed systems to serve millions of concurrent users, and mentoring junior engineers to grow their craft. My core technical expertise spans Python, React, Node.js, and cloud-native architectures — skills directly aligned with the requirements outlined in your job description.

What excites me most about ${company} is your commitment to engineering excellence and the opportunity to work on products that matter at scale. I thrive in collaborative, fast-paced environments where curiosity and ownership are celebrated.

I would love the opportunity to discuss how my background and enthusiasm can drive meaningful impact at ${company}. Thank you for considering my application — I look forward to the conversation.

Warm regards,

Alex Johnson
alex.johnson@email.com | linkedin.com/in/alexjohnson | github.com/alexjohnson`;
}

const TIPS = [
  { icon: '🎯', title: 'Tailor Every Letter', body: 'Customise each letter for the specific company and role. Generic letters have a 3× lower response rate.' },
  { icon: '📊', title: 'Use Numbers', body: 'Quantify achievements wherever possible. "Improved performance by 40%" beats "improved performance".' },
  { icon: '✂️', title: 'Keep It Concise', body: 'Aim for 250–350 words. Hiring managers spend an average of 7 seconds on first read.' },
];

export default function CoverLetter() {
  const [company, setCompany] = useState('');
  const [role, setRole]       = useState('');
  const [jd, setJd]           = useState('');
  const [tone, setTone]       = useState('Professional');
  const [letter, setLetter]   = useState('');
  const [loading, setLoading] = useState(false);
  const [error, setError]     = useState('');
  const [copied, setCopied]   = useState(false);
  const [editing, setEditing] = useState(false);

  async function generate() {
    if (!company.trim() || !role.trim()) { setError('Please fill in Company and Role.'); return; }
    setError(''); setLoading(true); setLetter('');
    try {
      const r = await fetch('/api/cover-letter', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ company, role, job_description: jd, tone }),
      });
      if (!r.ok) throw new Error(r.status);
      const d = await r.json();
      setLetter(d.cover_letter || '');
    } catch {
      setLetter(buildMockLetter(company, role));
    } finally {
      setLoading(false);
    }
  }

  function copyLetter() {
    navigator.clipboard.writeText(letter).then(() => {
      setCopied(true); setTimeout(() => setCopied(false), 2000);
    });
  }

  function downloadTxt() {
    const blob = new Blob([letter], { type: 'text/plain' });
    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    a.download = `cover_letter_${company.replace(/\s+/g, '_')}.txt`;
    a.click();
  }

  return (
    <Layout>
      <SectionHeader icon="✉️" title="Cover Letter Generator" />
      <p style={{ color: 'var(--text-muted)', marginBottom: '1.75rem', fontSize: '.92rem', maxWidth: 560 }}>
        Generate a professional, tailored cover letter in seconds — powered by IBM watsonx AI.
      </p>

      {/* ── Split screen layout ──────────────────────── */}
      <div className="col-2" style={{ alignItems: 'start', gap: '1.5rem' }}>
        {/* ── Left: Prompt ──────────────────────────── */}
        <div className="card">
          <div style={{ display: 'flex', alignItems: 'center', gap: '.6rem', marginBottom: '1.5rem' }}>
            <div style={{ width: 32, height: 32, background: 'var(--ibm-blue-soft)', borderRadius: 8, display: 'flex', alignItems: 'center', justifyContent: 'center' }}>✏️</div>
            <h3 style={{ fontSize: '.95rem', fontWeight: 600, color: 'var(--text-primary)' }}>Letter Details</h3>
          </div>

          <div className="form-group">
            <label>🏢 Company Name</label>
            <input
              className="form-control"
              placeholder="e.g. Stripe, Google, IBM…"
              value={company}
              onChange={e => setCompany(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label>🎯 Role / Position</label>
            <input
              className="form-control"
              placeholder="e.g. Senior Software Engineer"
              value={role}
              onChange={e => setRole(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label>📋 Job Description <span style={{ color: 'var(--text-muted)', fontWeight: 400 }}>(optional)</span></label>
            <textarea
              className="form-control"
              rows={6}
              placeholder="Paste the job description here for a more tailored letter…"
              value={jd}
              onChange={e => setJd(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label>📢 Writing Tone</label>
            <select className="form-control" value={tone} onChange={e => setTone(e.target.value)}>
              {['Formal', 'Professional', 'Enthusiastic', 'Bold'].map(t => <option key={t}>{t}</option>)}
            </select>
          </div>

          {error && <div className="warn-box" style={{ marginBottom: '1rem' }}>{error}</div>}

          <button
            className="btn btn-full"
            style={{ borderRadius: 12, padding: '.75rem' }}
            onClick={generate}
            disabled={loading}
          >
            {loading ? (
              <>
                <span style={{ display: 'inline-block', width: 14, height: 14, border: '2px solid rgba(255,255,255,0.3)', borderTopColor: '#fff', borderRadius: '50%', animation: 'spin .6s linear infinite' }} />
                Generating…
              </>
            ) : (
              <>✨ Generate Cover Letter</>
            )}
          </button>

          {/* Action buttons when letter exists */}
          {letter && !loading && (
            <div className="col-3" style={{ marginTop: '.85rem', gap: '.5rem' }}>
              <button className="btn btn-secondary btn-sm" style={{ borderRadius: 8 }} onClick={copyLetter}>
                {copied ? '✓ Copied!' : '📋 Copy'}
              </button>
              <button className="btn btn-secondary btn-sm" style={{ borderRadius: 8 }} onClick={downloadTxt}>
                💾 Download
              </button>
              <button className="btn btn-ghost btn-sm" style={{ borderRadius: 8 }} onClick={() => setEditing(!editing)}>
                {editing ? '💾 Done' : '✏️ Edit'}
              </button>
            </div>
          )}
        </div>

        {/* ── Right: Generated letter ──────────────── */}
        <div>
          {loading && <Loading message="IBM watsonx AI is crafting your cover letter…" />}

          {!loading && !letter && (
            <div style={{
              textAlign: 'center', padding: '5rem 2rem',
              background: 'rgba(255,255,255,0.02)',
              border: '1px solid var(--glass-border)',
              borderRadius: 20,
              height: '100%',
              minHeight: 400,
              display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center',
            }}>
              <div style={{
                width: 80, height: 80, borderRadius: '50%',
                background: 'rgba(255,255,255,0.04)',
                display: 'flex', alignItems: 'center', justifyContent: 'center',
                margin: '0 auto 1.25rem',
                fontSize: '2rem',
              }}>✉️</div>
              <div style={{ fontWeight: 600, color: 'var(--text-secondary)', marginBottom: '.4rem' }}>
                Your cover letter will appear here
              </div>
              <div style={{ fontSize: '.875rem', color: 'var(--text-muted)' }}>
                Fill in the form on the left and click Generate.
              </div>
            </div>
          )}

          {letter && !loading && (
            <div style={{ animation: 'slideUp .4s cubic-bezier(0.16,1,0.3,1)' }}>
              <div className="success-box" style={{ marginBottom: '1rem', display: 'flex', alignItems: 'center', gap: '.6rem' }}>
                <svg viewBox="0 0 24 24" width="16" height="16" fill="#34d399">
                  <path d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"/>
                </svg>
                Cover letter generated for <strong>{company}</strong>
                <span style={{ color: 'var(--text-muted)', marginLeft: 'auto' }}>{letter.split(/\s+/).length} words</span>
              </div>

              {editing ? (
                <textarea
                  className="form-control"
                  rows={20}
                  style={{ fontSize: '.875rem', lineHeight: 1.8, fontFamily: 'IBM Plex Sans, sans-serif' }}
                  value={letter}
                  onChange={e => setLetter(e.target.value)}
                />
              ) : (
                <div className="letter-box">{letter}</div>
              )}
            </div>
          )}
        </div>
      </div>

      {/* ── Tips ──────────────────────────────────────── */}
      <hr className="divider" />
      <SectionHeader icon="💡" title="Cover Letter Tips" />
      <div className="col-3">
        {TIPS.map(({ icon, title, body }) => (
          <div key={title} className="card" style={{ borderTop: '2px solid var(--ibm-blue)' }}>
            <div style={{ fontSize: '1.4rem', marginBottom: '.6rem' }}>{icon}</div>
            <div style={{ fontWeight: 600, marginBottom: '.4rem', color: 'var(--text-primary)', fontSize: '.9rem' }}>{title}</div>
            <div style={{ fontSize: '.82rem', color: 'var(--text-muted)', lineHeight: 1.6 }}>{body}</div>
          </div>
        ))}
      </div>
    </Layout>
  );
}
